chain/neutrino.go

```diff
@@ -335,55 +386,65 @@ func (s *NeutrinoClient) pollCFilter(hash *chainhash.Hash) (*gcs.Filter, error)
 }
 
 // Rescan replicates the RPC client's Rescan command.
-func (s *NeutrinoClient) Rescan(startHash *chainhash.Hash, addrs []btcutil.Address,
-	outPoints map[wire.OutPoint]btcutil.Address) error {
-
+//
+// Rescan determines the current best block and checks to see if a
+// NeutrinoRescanner exists.  If a rescanner exists, the rescanner is shutdown
+// gracefully and a new NeutrinoRescanner is created. Rescan sends a
+// RescanFinished notification if the startHash matches the hash of the
+// best block.
+func (s *NeutrinoClient) Rescan(
+	startHash *chainhash.Hash,
+	addrs []btcutil.Address,
+	outPoints map[wire.OutPoint]btcutil.Address,
+) error {
 	s.clientMtx.Lock()
+	defer s.clientMtx.Unlock()
+
 	if !s.started {
-		s.clientMtx.Unlock()
 		return fmt.Errorf("can't do a rescan when the chain client " +
 			"is not started")
 	}
-	if s.scanning {
-		// Restart the rescan by killing the existing rescan.
-		close(s.rescanQuit)
-		rescan := s.rescan
-		s.clientMtx.Unlock()
-		rescan.WaitForShutdown()
-		s.clientMtx.Lock()
-		s.rescan = nil
-		s.rescanErr = nil
+
+	inputsToWatch, err := toInputsToWatch(outPoints)
+	if err != nil {
+		return err
 	}
-	s.rescanQuit = make(chan struct{})
-	s.scanning = true
-	s.finished = false
-	s.lastProgressSent = false
-	s.lastFilteredBlockHeader = nil
-	s.isRescan = true
-	s.clientMtx.Unlock()
 
 	bestBlock, err := s.CS.BestBlock()
 	if err != nil {
 		return fmt.Errorf("can't get chain service's best block: %s", err)
 	}
+
 	header, err := s.CS.GetBlockHeader(&bestBlock.Hash)
 	if err != nil {
 		return fmt.Errorf("can't get block header for hash %v: %s",
 			bestBlock.Hash, err)
 	}
 
-	// If the wallet is already fully caught up, or the rescan has started
-	// with state that indicates a "fresh" wallet, we'll send a
-	// notification indicating the rescan has "finished".
-	if header.BlockHash() == *startHash {
-		s.clientMtx.Lock()
-		s.finished = true
-		rescanQuit := s.rescanQuit
-		s.clientMtx.Unlock()
+	select {
+	case rescanner := <-s.rescannerCh:
```

Comment:
All these changes to the rescan logic are very hard to reason about for a reviewer. Is there any way to split this big commit up into small pieces that logically build upon each other to help understand all the individual changes?